design/automatic-network-configuration.md

```diff
@@ -0,0 +1,485 @@
+// Port contains the information of its own capabilities, connected network device
+// and specific port number of the device.
+type Port struct {
+    ConPortID    uint
+    ConDeviceRef struct {
+        Name      string
+        Namespace string
+        Kind      string
+    }
+    Capability struct {
+        Speed uint
+        SmartNic bool
+    }
+}
+```
+
+![design compose](images/automatic_network_configuration_design_compose.png)
+
+### workflow
+
+Create CR related to BMH's network:
+
+* Administrator creates the corresponding `Device` CR according to the
+  specific network device.
+* Administrator creates a `BareMetalHost` CR and fills in the content in `bmh.spec.ports`
+  according to the actual network connection.
+
+Create Metal3Machine:
+
+* User creates `NetworkConfiguration` CR.
+* User specifies the `NetworkConfiguration` CR used when creating `Metal3Machine`
+  CR.
+* CAPM3 selects a group of BMHs according to the original logic.
+  If the `networkConfiguration` field of Metal3Machine is not empty,
+  call the `filter()` method, select and filter the BMH suitable for the network
+  configuration from the currently available BMHs, and then select one BMH
+  according to the original method.
+* Before provisioning, if the machine's network configuration field is not empty,
+  CAPM3 calls the `configureNetwork()` method. In this method, CAPM3 triggers the
+  NetworkBinding Controller to configure the BareMetal network by creating
+  NetworkBinding CRs. In addition, CAPM3 also needs to add the `Metal3Machine` to
+  the `.metaData.OwnerReferences` of the `NetworkBinding` CR to ensure that when
+  the `Metal3Machine` is deleted, the corresponding `NetworkBinding` CR will also
+  get deleted, which triggers the NetworkBinding Controller to clear the network
+  configuration corresponding to BareMetal.
+* NetworkBinding controller starts to configure the network.
+* `configureNetwork()` method polls the `status.state` field of the
+  `NetworkBinding` CR to determine the result of the network configuration.
+* After the network configuration is successful, CAPM3 continues to provision BMH.
+
+Remove Metal3Machine:
+
+* User deletes the `Metal3Machine` CR. Since the OwnerReferences of the
+  `NetworkBinding` CR is the `Metal3Machine` CR, k8s will also delete the
+  `NetworkBinding` CR.
+* The NetworkBinding Controller detects that the `NetworkBinding` CR needs to be
+  deleted and starts to clear the network configuration.
+* CAPM3 calls the `deconfigureNetwork()` method to detect whether the
+  `NetworkBinding` CR has been deleted, and returns the result.
+* If the network configuration is deleted successfully, continue to
+  deprovision the BMH.
+
+![timing diagram](images/automatic_network_configuration_timing_diagram.png)
+
+### Changes to Current API
+
+#### Metal3Machine CRD
+
+```yaml
+spec:
+  # Specify the network configuration that metal3Machine needs to use
+  networkConfigurationRef:
+  - name: nc1
+    namespace: default
```

Comment:
I am still unclear on how, as a user, you specify which port of the BMH should have which configuration. How do you pick the switch port on which to apply this configuration ? In your proposal, the metal3Machine seems to be the only point where you can bring together the switch configuration (from linking it to the BMH port) and the cloud-init data that you will pass to the BMH. (It could also be done at the BMH level if there was no CAPI in the picture).

The user will need to specify the network configuration to be applied via cloud-init on the host. How can this be mapped with the configuration applied on the switch. The goal is to ensure that if you configure interface 1 of the OS, the proper configuration is applied on the switch in a deterministic manner to ensure both configurations match.

Reply:
The user cannot specify which interafce or which port to configure. In our opinion, the user does not know which BMH will be selected and which switchport is connected to this BMH.
The user just declares which network configuration is required for the created Metal3Machine, and then the configuration of the specific port is performed by the operator.

Reply:
That's a problem. How do you configure properly the network configuration for the OS via cloud-init if you don't know which interface is configured with that configuration ?

Reply:
After the user declares the networkConfiguration used by Metal3Machine, CAPM3 selects the appropriate BMH according to the configuration requirements, then selects the network card that meets the network configuration requirements, and then creates NetworkBounding according to the port connected to the network card to trigger the network configuration.

Reply:
That is for configuring the switch. But how do you map that with the network configuration that needs to be done on the host by cloud-init for example ? How do you know what interface is connected to which vlan once you are on the host running cloud-init ?